### PR TITLE
Support 0-d tensors in CUDA MultiLabelMarginCriterion.

### PR DESCRIPTION
**Pull request by gchanan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30773 [BC-BREAKING] change index_select scalar_check to retain dimensionality of input.
* #30772 Turn off scalar_checks for SpatialDepthwiseConvolution and SpatialConvolutionMM.
* #30753 Move scalar_check from codegen to code in MultiLabelMarginCriterion.
* #30748 [BC-Breaking] Fix scalar check of MultiLabelMarginLoss.
* #30733 Fix a CUDA memory leak in MultiLabelMarginCriterion error checking.
* #30728 Turn off scalar_check for is_target for MultiLabelMarginCriterion, which is handled correctly in code.
* **#30727 Support 0-d tensors in CUDA MultiLabelMarginCriterion.**
* #30726 MultiMarginCriterion: move scalar_check to code instead of codegen.

It is already supported in CPU and is pretty easy to add for consistency.

Differential Revision: [D18808768](https://our.internmc.facebook.com/intern/diff/D18808768)

### CODE REVIEW

**Review by albanD — CHANGES_REQUESTED** [review]
From the [doc](https://pytorch.org/docs/stable/nn.html#torch.nn.MultiLabelMarginLoss) the only acceptable inputs are 1D or 2D. I think we should do:
- Change the doc to allow 0D/0D pair. But then check that `input.dim() == target.dim()`.
- Change the code to forbid any 0D input or target.
